### PR TITLE
[JAV-429] modify the startup policy of SDK

### PR DESCRIPTION
**Pull request by weichao666**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://https://servicecomb.atlassian.net/projects/JAV/issues) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[JAV-XXX] Fixes bug in ApproximateQuantiles`, where you replace `JAV-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: service-registry/src/main/java/io/servicecomb/serviceregistry/api/request/CreateSchemaRequest.java [hunk 1]
```diff
@@ -20,11 +20,22 @@
 public class CreateSchemaRequest {
   private String schema;
 
+  private String summary;
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/api/request/CreateSchemaRequest.java]: signature is better?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/client/http/ServiceRegistryClientImpl.java [hunk 2]
```diff
@@ -293,6 +321,56 @@ public String getSchema(String microserviceId, String schemaId) {
...
+        });
+    try {
+      countDownLatch.await();
+    } catch (Exception e) {
+      LOGGER.error("query schemasummary exist {}/{} failed",
+          microserviceId,
+          schemaId,
+          e);
+    }
+    if (getSchemaResponse.getSchemaSummary() != null) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/client/http/ServiceRegistryClientImpl.java]: return getSchemaResponse.getSchemaSummary()

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 3]
```diff
@@ -113,8 +116,8 @@ private boolean checkSchemaIdSet() {
     schemaIdSetMatch = existSchemas.equals(localSchemas);
 
     if (!schemaIdSetMatch) {
-      LOGGER.error(
-          "SchemaIds is different between local and service center. Please change microservice version. "
+      LOGGER.warn(
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: product mode, it's a error, why change to warn?
and must stop boot.

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 4]
```diff
@@ -113,8 +116,8 @@ private boolean checkSchemaIdSet() {
...
+          "SchemaIds is different between local and service center. "
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: old suggestion is change version.
what's the new suggestion?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 5]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
       boolean exists = srClient.isSchemaExist(microservice.getServiceId(), schemaId);
       LOGGER.info("schemaId {} exists {}", schemaId, exists);
       if (!exists) {
-        if (!srClient.registerSchema(microservice.getServiceId(), schemaId, content)) {
-          return false;
+        if (serviceCenterEnvIsDev()) {
+            if (!srClient.registerSchema(microservice.getServiceId(), schemaId, content)) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: product mode, will not register schema?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 6]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
...
+          }
         }
       }
     }
 
     return true;
   }
+
+  private boolean serviceCenterEnvIsDev() {
+    if (Const.SERVICECENTER_RUNMODE_DEV.equals(srClient.getServiceCenterEnvironment().getRunMode())) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: always query from remote?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 7]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
...
+      LOGGER.warn("The current servicecenter environment runmode is {}", Const.SERVICECENTER_RUNMODE_DEV);
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: if have 100 schemas, then print 100 warnings?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 8]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
...
+        if (serviceCenterEnvIsDev()) {
+            if (!srClient.registerSchema(microservice.getServiceId(), schemaId, content)) {
+                return false;
+            }
+        }
+      } else {
+        //query if schema exist, if not exist, register schema&summary in SC
+        String serviceId = microservice.getServiceId();
+        String schema = srClient.getSchema(serviceId, schemaId);
+        if (null == schema) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: we already make sure schema is exist, but here we can not get it
can we override remote schema directly?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 9]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
       boolean exists = srClient.isSchemaExist(microservice.getServiceId(), schemaId);
       LOGGER.info("schemaId {} exists {}", schemaId, exists);
       if (!exists) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: following logic is too complex, too deep determine
extract methods.

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 10]
```diff
@@ -143,12 +146,53 @@ private boolean registerSchemas() {
...
+        String serviceId = microservice.getServiceId();
+        String schema = srClient.getSchema(serviceId, schemaId);
+        if (null == schema) {
+          LOGGER.warn("schemacontent {}/{} does not exist.", serviceId, schemaId);
+          if (!srClient.registerSchema(serviceId, schemaId, content)) {
+            return false;
+          }
+        } else {
+          //query if schemaSummary exist, if not exist, register schema&summary in SC
+          String schemaSummary = srClient.getSchemaSummary(serviceId, schemaId);
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: i remember we should merge getSchema and getSchemaSummary to one remote invoke

File: service-registry/src/main/java/io/servicecomb/serviceregistry/api/request/CreateSchemaRequest.java (hunk 1 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/api/request/CreateSchemaRequest.java]: done

File: service-registry/src/main/java/io/servicecomb/serviceregistry/client/http/ServiceRegistryClientImpl.java (hunk 2 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/client/http/ServiceRegistryClientImpl.java]: done

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 3 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 4 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done, still use the old suggestion

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 5 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done, if not exist, just register schema

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 6 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done, add a boolean scEnvIsDev to resolve

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 7 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done,now print only once

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 8 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done，there is no need to query schema, if schemaId exist, just query schemaSummary

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 9 above)
Comment (weichao666) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: done

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java (hunk 3 above)
Comment (wujimin) [reply service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: ???
product mode , if check failed, must stop boot?

i can not find the logic.

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 11]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean serviceCenterEnvIsDev() {
+    if (Const.SERVICECENTER_RUNMODE_DEV.equals(srClient.getServiceCenterEnvironment().getRunMode())) {
+      LOGGER.warn("The current servicecenter environment runmode is {}", Const.SERVICECENTER_RUNMODE_DEV);
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: print the const value?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 12]
```diff
@@ -122,6 +127,9 @@ private boolean checkSchemaIdSet() {
           microservice.getVersion(),
           localSchemas,
           existSchemas);
+      if (!serviceCenterEnvIsDev()) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: mode determined by SC?
i remember changed design that determined by microservice?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 13]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
     return true;
   }
+
+  private boolean checkSchemaSummary(String schemaId, String content) {
+    //query if schemaSummary exist, if not exist, register schema&summary in SC
+    String serviceId = microservice.getServiceId();
+    String schemaSummary = srClient.getSchemaSummary(serviceId, schemaId);
+    if (null == schemaSummary) {
+      return srClient.registerSchema(serviceId, schemaId, content);
+    } else {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: already return
no need to write "else"

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 14]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
+      //compare to localschema, if don't match,when SC runMode is dev,register schema&summary in SC
+      String localSummary = Hashing.sha256().newHasher().putString(content, StandardCharsets.UTF_8).hash().toString();
+      if (!localSummary.equals(schemaSummary)) {
+        if (scEnvIsDev || serviceCenterEnvIsDev()) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: scEnvIsDev || serviceCenterEnvIsDev()????
what's you want?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 15]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
+          return srClient.registerSchema(serviceId, schemaId, content);
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: why "check" function, include register logic?

**Comment by wujimin** [comment]
too many problems.
please review it yourself carefully.

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 16]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
+  private boolean checkSchemaSummary(String schemaId, String content) {
+    //query if schemaSummary exist, if not exist, register schema&summary in SC
+    String serviceId = microservice.getServiceId();
+    String schemaSummary = srClient.getSchemaSummary(serviceId, schemaId);
+    if (null == schemaSummary) {
+      return srClient.registerSchema(serviceId, schemaId, content);
+    } else {
+      //compare to localschema, if don't match,when SC runMode is dev,register schema&summary in SC
+      String localSummary = Hashing.sha256().newHasher().putString(content, StandardCharsets.UTF_8).hash().toString();
+      if (!localSummary.equals(schemaSummary)) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]:
```
if (localSummary.equals(schemaSummary)) {
  return true;
}

return scEnvIsDev;
```
do you think this is simpler?

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 17]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
         if (!srClient.registerSchema(microservice.getServiceId(), schemaId, content)) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: return srClient.registerSchema(microservice.getServiceId(), schemaId, content);

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 18]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
           return false;
         }
+      } else {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: no need to write "else"

File: service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java [hunk 19]
```diff
@@ -146,9 +154,43 @@ private boolean registerSchemas() {
...
+        if (!checkSchemaSummary(schemaId, content)) {
```
Comment (wujimin) [inline service-registry/src/main/java/io/servicecomb/serviceregistry/task/MicroserviceRegisterTask.java]: all existing schema, must invoke SC two times:
1.check if exists
2.check signature

can we merge it to be a single invoke?

and you already modify name to signature, why name summary here?

**Pull request closed without merging**